tests/unittests/CoreGraphics.drawing/CGContextDrawing_ImageMaskTests.cpp

```diff
@@ -0,0 +1,314 @@
+    }
+
+    return maskImage.release();
+}
+
+auto __MakeSquareMaskGenerator(CGAffineTransform transform) {
+    // All this work to create a square/diamond with a diamond hole in the middle.
+    return [transform](CGContextRef context, CGSize size, ClippingType clipType) {
+        if (clipType == ClippingTypeMask) {
+            // Since it's a mask, flood it with white.
+            CGContextSetRGBFillColor(context, 1.0, 1.0, 1.0, 1.0);
+            CGContextFillRect(context, { CGPointZero, size });
+        }
+
+        CGContextTranslateCTM(context, size.width / 2.f, size.height / 2.f);
+        CGContextConcatCTM(context, transform);
+        CGContextTranslateCTM(context, -(size.width / 2.f), -(size.height / 2.f));
+
+        CGContextBeginPath(context);
+
+        // Add a set of nested diamonds and clip to them. They should be anchored at 1/4 and 2/3 of the way through the context,
+        // respectively.
+        CGContextAddRect(context, { size.width / 4.f, size.height / 4.f, size.width / 2.f, size.height / 2.f });
+        CGContextAddRect(context, { size.width * .75 / 2.f, size.height * .75 / 2.f, size.width / 4.f, size.height / 4.f });
+        CGContextEOClip(context);
+
+        CGContextSetRGBFillColor(context, 0.5, 0.5, 0.5, clipType == ClippingTypeAlpha ? 0.5 : 1.0);
+        CGContextFillRect(context, { CGPointZero, size });
+
+        // Add an additional full-brightness region (the bottom tip of the diamond.)
+        CGContextSaveGState(context);
+        CGContextAddRect(context, { CGPointZero, size.width / 2.f, size.height / 2.f });
+        CGContextClip(context);
+
+        CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1.0);
+        CGContextFillRect(context, { CGPointZero, size });
+    };
+}
+
+void __GradientMaskGenerator(CGContextRef context, CGSize size, ClippingType clipType) {
+    for (size_t i = 0; i < 256; ++i) {
+        float val = (float)i / 255.f;
+        CGContextSetRGBStrokeColor(context, val, val, val, clipType == ClippingTypeAlpha ? (1.0 - val) : 1.0);
+        CGPoint line[]{
+            { i + .5, 0 }, { i + .5, size.height },
+        };
+        CGContextStrokeLineSegments(context, line, 2);
+    }
+}
+
+class CGContextClipping : public ::testing::DrawTest,
+                          public ::testing::WithParamInterface<::testing::tuple<ClippingShape, ClippingType, CGSize>> {
+private:
+    static std::vector<std::tuple<ClippingShape, CGSize, std::function<void(CGContextRef, CGSize, ClippingType)>>> s_maskGenerators;
+
+protected:
+    CFStringRef CreateOutputFilename() {
+        const ::testing::TestInfo* const test_info = ::testing::UnitTest::GetInstance()->current_test_info();
+        ClippingShape shape = ::testing::get<0>(GetParam());
+        ClippingType type = ::testing::get<1>(GetParam());
+        const char* shapeName = "unk";
+        const char* typeName;
+        switch (shape) {
+            case ClippingShapeDiamond:
+                shapeName = "di";
```

Comment:
it's glorious. It's not a `char* const` afterall!

---
In reply to: [92094081](https://github.com/Microsoft/WinObjC/pull/1528#discussion_r92094081) [](ancestors = 92094081)